app/assets/javascripts/discourse/app/lib/singleton.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * @decorator
+ * Ensures only one instance of a class exists and provides global access to it.
+ *
+ * @example
+ * ```
+ * @singleton
+ * class UserSettings {
+ *   theme = 'light';
+ *
+ *   toggleTheme() {
+ *     this.theme = this.theme === 'light' ? 'dark' : 'light';
+ *   }
+ * }
+ *
+ * // Get the singleton instance
+ * const settings = UserSettings.current();
+ *
+ * // Access properties
+ * UserSettings.currentProp('theme'); // 'light'
+ * UserSettings.currentProp('theme', 'dark'); // sets and returns 'dark'
+ *
+ * // Multiple calls return the same instance
+ * UserSettings.current() === UserSettings.current(); // true
+ * ```
+ */
+export default function singleton(target) {
+  return class SingletonClass extends target {
```

Comment:
The downside of this is that the resultant class will always be named 'SingletonClass' when viewed in a console:

<img width="551" alt="SCR-20241230-otkb" src="https://github.com/user-attachments/assets/0c3a24d7-f025-4254-b9fc-8d20d0b71695" />

So I wonder whether we should be mutating the class rather than returning a child class? 🤔